src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -65,12 +65,47 @@ private function completeDefinition($id, Definition $definition)
 
         $this->container->addClassResource($reflectionClass);
 
-        if (!$constructor = $reflectionClass->getConstructor()) {
+        if ($constructor = $reflectionClass->getConstructor()) {
+            $this->autowireMethod($id, $definition, $constructor, true);
+        }
+
+        $methodsCalled = array();
+        foreach ($definition->getMethodCalls() as $methodCall) {
+            $methodsCalled[$methodCall[0]] = true;
+        }
+
+        foreach ($reflectionClass->getMethods(\ReflectionMethod::IS_PUBLIC) as $reflectionMethod) {
+            $name = $reflectionMethod->getName();
+            if (isset($methodsCalled[$name]) || $reflectionMethod->isStatic() || 0 !== strpos($name, 'set')) {
+                continue;
+            }
+
+            $this->autowireMethod($id, $definition, $reflectionMethod, false);
+        }
+    }
+
+    /**
+     * Autowires the constructor or a setter.
+     *
+     * @param string            $id
+     * @param Definition        $definition
+     * @param \ReflectionMethod $reflectionMethod
+     * @param bool              $constructor
+     *
+     * @throws RuntimeException
+     */
+    private function autowireMethod($id, Definition $definition, \ReflectionMethod $reflectionMethod, $constructor)
```

Comment:
What do you think to rename ``$constructor`` to ``$isConstructor`` to be clearer?